Add BaseKnnVectorsFormatTestCase.testRecall() and fix old codecs

msokolov opened the pull request:
While exploring some recall-related failures in another PR I went looking for a unit test that checks HNSW/KNN recall and couldn't find any. I think we used to have one but maybe we removed it because it was flaky? But we really do need such a test since it is possible to make changes that preserve all the formal properties of the codecs and queries yet destroy recall. I thought if we can create such a test with known data and vectors it would be more predictable than one using random data, so I made one, and it uncovered a couple of bugs: 

In Lucene90HnswVectorsReader we messed up (removed) ord-to-doc mappings so we were returning vector ords instead of docids in search results. I guess this would have totally borked back-compat for Lucene90 indexes. Probably there are none in the wild, and this was never noticed?

In Lucene91RWFormat (used only for back-compat testing) we messed up diversity check so we were producing bad graphs.

This PR fixes these things and adds the new test

Commit:
Add BaseKnnVectorsFormatTestCase.testRecall() and fix map ord to doc in Lucene90HnswVectorsReader

Commit:
Add BaseKnnVectorsFormatTestCase.testRecall() and fix map ord to doc in Lucene90HnswVectorsReader

msokolov commented:
I wonder whether we should backport the fixes to the `Lucene90HnswVectorsReader`? I tend to think we ought to, although the usage might be tiny to nonexistent

benwtrent commented on lucene/backward-codecs/src/java/org/apache/lucene/backward_codecs/lucene90/Lucene90HnswVectorsReader.java:
```diff
@@ -260,7 +260,7 @@ public void search(String field, float[] target, KnnCollector knnCollector, Bits
       float minSimilarity = results.topScore();
       results.pop();
-      knnCollector.collect(node, minSimilarity);
+      knnCollector.collect(vectorValues.ordToDoc(node), minSimilarity);
```
Since Lucene90 didn't support sparse vector values, I am not sure this is strictly necessary. But I can understand it from a consistency standpoint.

msokolov replied:
Oh, that's a relief! I couldn't remember if we had that or not. At any rate it is possible to create a sparse 90 index in tests now.

mikemccand replied:
> Oh, that's a relief! I couldn't remember if we had that or not. At any rate it is possible to create a sparse 90 index in tests now.

This is odd?  When a Codec format is replaced, we move the "read only" part to `backwards-codecs` module, and the full read/write original codec to `test-framework` so that unit tests can produce a 9.0 index and run modern tests (even new unit tests added since 9.0) against such "old" indices.

But that read/write codec format should not be able to produce an index that the original 9.0 (core) format was not able to produce.  They should be the same original code ... so maybe sparse vector values were in fact writable in 9.0 (by accident?)?

benwtrent reviewed (commented):
I think we should have a test like this. I wonder if its better to just grab 100-500 sift vectors and put them in a file though.

msokolov commented on lucene/test-framework/src/java/org/apache/lucene/tests/index/BaseKnnVectorsFormatTestCase.java:
```diff
@@ -1906,4 +1916,122 @@ public void testMismatchedFields() throws Exception {
+   * gross failures only, not to represent the true expected recall.
+   */
+  public void testRecall() throws IOException {
+    VectorSimilarityFunction vectorSimilarityFunction = VectorSimilarityFunction.EUCLIDEAN;
```
agreed, we want to test all the things

msokolov commented on lucene/test-framework/src/java/org/apache/lucene/tests/index/BaseKnnVectorsFormatTestCase.java:
```diff
@@ -1906,4 +1916,122 @@ public void testMismatchedFields() throws Exception {
+      // pass match-all "filter" to force full traversal, bypassing graph
+      KnnFloatVectorQuery exactQuery =
+          new KnnFloatVectorQuery("field", queryEmbedding, 1000, new MatchAllDocsQuery());
+      // indexed 421 lines from LICENSE.txt
+      // indexed 157 lines from NOTICE.txt
```
lemme see if I can find bugs with the "dictionary" PR? We can always beef this up with more realistic data

msokolov commented on lucene/test-framework/src/java/org/apache/lucene/tests/index/BaseKnnVectorsFormatTestCase.java:
```diff
@@ -1906,4 +1916,122 @@ public void testMismatchedFields() throws Exception {
+          ++recalled;
+        }
+      }
+      assertTrue("recall should be at least 5/10, got " + recalled, recalled >= 5);
```
I think it could be tricky to be very precise given the range of codec options. I guess we could specialize per codec?

benwtrent replied:
I think having an `assertAvgRecall` in the base class that can be overridden would be really nice.

msokolov reviewed (commented):
thanks for review!

benwtrent commented on lucene/test-framework/src/java/org/apache/lucene/tests/index/BaseKnnVectorsFormatTestCase.java:
```diff
@@ -1906,4 +1916,122 @@ public void testMismatchedFields() throws Exception {
+      computeLineEmbedding(queryString, queryEmbedding);
+      // computeLineEmbedding("   END OF TERMS AND CONDITIONS", queryEmbedding);
+      // pass match-all "filter" to force full traversal, bypassing graph
+      KnnFloatVectorQuery exactQuery =
+          new KnnFloatVectorQuery("field", queryEmbedding, 1000, new MatchAllDocsQuery());
```
Also, I think for more consistent runs, we may want to have multiple query embeddings that we test with and gather `min` `max` and `avg` recalls. But this can be a further refinement on this work.

I just think having a single query might be very flaky in the long run.

Commit:
handle stray prints

Commit:
test all similarities and more queries

Commit:
refactor assertRecall method

msokolov commented:
@benwtrent I think I addressed all your comments except adding binary vectors. I think as long as the vectors are not too degenerate and always the same the test purpose is satisfied, but I'm not averse to replacing these janky ones with more "scientific" ones either, I just don't have the handy.

This test seems to pass on multiple runs and it is exposing problems with the dictionary refactor I'm working on, so I think it's helpful

benwtrent reviewed (approved):
LGTM! This is a nice sanity check for abhorrent behavior :)

msokolov commented:
> Since Lucene90 didn't support sparse vector values, I am not sure this is strictly necessary. But I can understand it from a consistency standpoint.

After reflection, I don't think this is true. We always supported sparse vector indexes - I don't see how we could have avoided it, really. It seems to me this bug was introduced  [here](https://github.com/apache/lucene/commit/a65cf8960a1057d98126256d1610292ad5c8f1b3#diff-b75a9bbd95dd9708267528a08f3d0b1093f74feb179f03f16f8dfb1857930077L271) and it was released as part of 9.8, meaning that releases 9.8 and later in the 9.x series will be able to read indexes produced in 9.0 but will give meaningless results for HNSW searches over those indexes. This seems like something we maybe ought to make the user community aware of

msokolov commented:
ok something like this:

Dear Lucene user community,

We recently uncovered a backwards compatibility bug that affects indexes created with version 9.0 containing KNN vector fields. Versions 9.8 - 9.12 are unable to search vectors in such indexes correctly and will return incorrect results without raising any error. We think it's likely very few if any of you are using 9.0 indexes, but if you are, possible mitigation steps are:

* Upgrade to 10.0 or later, or
* Do not upgrade past 9.7, or
* If you must use an affected Lucene version (9.8-9.12) and you have 9.0-written indexes including KNN vector fields, you must recreate those indexes from source with your current Lucene version.

benwtrent commented:
@msokolov could we do a simpler patch for 9.12.1?

msokolov commented:
Yes, maybe we should -- I think it would be a one-liner

msokolov commented:
There is another upgrade path -- if you started with 9.0 and then "upgraded" your index by rewriting it (eg with IndexUpdater tool) via merge to 9.1-9.7 you could subsequently read the index with later versions. But this seemed kind of complex to explain for a case that probably doesn't exist.

benwtrent commented:
@mikemccand I have a PR open for this bug fix for 9.12. Will merge soon. 

Could you add a CHANGES entry in 9.12 for your bug fix for 9.12.1?

mikemccand commented:
> Could you add a CHANGES entry in 9.12 for your bug fix for 9.12.1?

Ahh yes sorry I will do that today!
